### ISSUE TITLE
Issues installing on CentOS

### ISSUE THREAD

**Issue by phil-grayson** [Roy-lab/Arboretum2.0#1]
After cloning the repository, I attempted to run `make` but received a number of errors involving ‘getpid'.  They are included below:
```
make
g++ Framework.C  SpeciesDistManager.C	SpeciesClusterManager.C	Expert.C Matrix.C GeneExpManager.C MappedOrthogroupReader.C     MappedOrthogroup.C GeneMap.C GeneTreeManager.C GammaManager.C Gamma.C GeneTree.C GeneNameMapper.C NewickReader.C BioNetwork.C  BioNode.C Gene.C  GeneManager.C  Graph.C  Interaction.C  InteractionManager.C  Node.C  Path.C  Protein.C  ProteinManager.C BFGSWrapper.C ClusterManager.C Kmeans.C  ExpertL.C MotifManager.C MotifRegressor.C Randomizer.C TestData.C  Distance.C HyperGeomPval.C common/Variable.C common/VariableManager.C common/Evidence.C common/EvidenceManager.C common/Error.C -I include -I common  -L lib -lgsl -lgslcblas  -g -o arboretum
SpeciesClusterManager.C: In member function ‘int SpeciesClusterManager::readSpeciesData(const char*)’:
SpeciesClusterManager.C:118:16: error: ‘getpid’ was not declared in this scope
   rseed=getpid();
                ^
SpeciesClusterManager.C: In member function ‘int SpeciesClusterManager::readSpeciesData(const char*, const char*)’:
SpeciesClusterManager.C:173:30: error: ‘getpid’ was not declared in this scope
                 rseed=getpid();
                              ^
GammaManager.C: In member function ‘int GammaManager::sampleAllClusterAssignments(std::map<int, std::map<std::basic_string<char>, int>*>&)’:
GammaManager.C:1848:19: error: ‘getpid’ was not declared in this scope
  int rseed=getpid();
                   ^
^Cmake: *** [condSpecLearner] Interrupt

```

**Comment by asiahpirani** [comment]
Thank you for your interest in our work.

I think this is an issue with compiler version. It was working in the older versions of gcc but in newer versions we need to add "#include <unistd.h>" for getpid function.

I think Sara updated the code to have this include yesterday. Can you clone the code again and see if you still get the error?

**Comment by phil-grayson** [comment]
Hi,

Thanks for the reply!  Yes.  I was in communication with Sara about this yesterday and she requested that I create the issue so that it’s documented.  I believe that she plans to respond to close it out.

All the best!

Phil

On May 13, 2020, at 11:56 AM, asiahpirani <notifications@github.com<mailto:notifications@github.com>> wrote:



Thank you for your interest in our work.

I think this is an issue with compiler version. It was working in the older versions of gcc but in newer versions we need to add "#include <unistd.h>" for getpid function.

I think Sara updated the code to have this include yesterday. Can you clone the code again and see if you still get the error?

**Comment by phil-grayson** [comment]
Thanks Ali! Yes, indeed, I’m just about to close that out now, and I made the update yesterday, which resolved the issue. Thanks again! Sara

**Comment by sknaack** [comment]
Hi Phil, 

I can readily update you now with a solution and I should be able to help you get it compiled readily today. There are two points. 

1. In a few instances there was an include necessary to smoothly compile the calls to the getpid() function in CentOS. In three files I added “#include "#include <unistd.h>” to address errors related to this function, specifically GammaManager.C, MotifRegressor.C and SpeciesClusterManager.C. With this update I was able to compile Arboretum from the repository without errors or warnings on one of our servers running CentOS. This alone may already largely help you. 

2. I did also make sure it was compiling in OS X by testing this on my laptop. With the above-mentioned change alone the included GSL library installation in the lib and include directories of the repository were (unsurprisingly), still not working for compilation. I did then install GSL from the source code at http://mirror.clarkson.edu/gnu/gsl/gsl-latest.tar.gz on my laptop to /usr/local/lib or /use/local/include path (the default path for this source code). This took just a few minor commands that I’ve indicated at the bottom of the README now. In the Makefile I also updated the LIBPATH and INCLPATH1 variables with these two paths to my local, OS X-compiled  GSL library installation, respectively. These paths are likely also found by default in OS X. With this change to the Makefile the compilation of Arboretum worked, albeit with a few warnings not generated in our general computing environment.

So if you download the current repository from https://github.com/Roy-lab/Arboretum2.0/archive/Version2.3.zip now, between this first point and addressing this second point for calling the GSL library in the Makefile (as needed for your environment), the compilation should proceed smoothly. Let me know if you find any additional concerns. It’s good to have some feedback. If you would be so kind as to enter this issue as a ticket on GitHub we’ll also document this there for any future users. Thanks!

Best wishes, 

Sara
